stip/stip-001-burn-election.md

```diff
@@ -0,0 +1,127 @@
+# Abstract
+
+This proposal describes a mechanism for single-leader election using
+burn-mining. Leader election is intimately connected to the concept of
+chain selection-- as leaders must be able to choose their chain tip, and
+that choice is impacted by network participants' rules for chain selection.
+
+# Motivation
+
+Leader election on the Stacks chain enables:
+
+1. Near instantaneous confirmation of Stacks transactions
+2. High validation throughput
+3. An open leadership set
+4. Opportunity to participate without mining hardware
+5. Ability to migrate to a separate burn chain in the future
+
+# Election
+
+The basic structure for leader election through burn-mining is that
+for some Stacks block _N_, the leader is selected via some function of
+that leader's total cryptocurrency burnt in a block _N'_ on the
+underlying burn chain (or over a window of blocks).
+
+In such a system, if a candidate _Alice_ wishes to be a leader of a
+Stacks block, she issues a burn transaction in the underlying burn
+chain. The network then uses cryptographic sortition to choose a
+leader in a verifiably random process, weighted by burn amounts.
+The block in which this burn transaction is broadcasted is known
+as the "election block" for Stacks block _N_.
+
+## Committing to a chain tip
+
+Leaders must be able to choose which chain tip they want to build off
+of. Otherwise, malicious leaders would be able to force subsequent
+leaders to follow their chains, even if they misbehaved by hiding
+transactions, broadcasting partial blocks, etc.
```

Comment:
Regarding malicious behavior on the leader's part pertaining to dropping transactions or not committing during their tenure, we would at least want the following to be true:

* If a leader commits _at all_ during their tenure, then they commit at least one transaction on-chain that the next leader can build off of.
* An honest leader should be able to recover from periods of inactivity without paying a penalty, but we only want leaders to be able to be able to do so if there were no on-chain commitments.

We can guarantee these if the leader must commit their Stacks block's coinbase transaction to the burn-chain.  To see why, consider these four possible outcomes when a leader L begins its tenure:

1. L commits to a microblock from the previous leader L_prev, and announces their microblocks in a timely fashion (the happy path).
2. L does not commit to a microblock during their tenure.
3. L commits to a valid microblock, but does not announce their microblocks.
4. L commits to a range of one or more leaderless blocks (i.e. the case where L_prev did not commit to a block during its tenure), and may or may not announce its microblocks.

In case 1, everything works out like we'd expect.

In case 2, L cannot broadcast anything. All nodes will ignore L's microblocks, and L will not receive a coinbase reward (i.e. this is equivalent to L opting to burn their coinbase). No other previous or future leaders' rewards are affected.  This discourages L from doing nothing, and L gains nothing by withholding transactions (see below).  L_next, the next leader, can recover in case 4.

In case 3, we need to make it so L has until the end of its tenure to broadcast microblocks.  If it doesn't, then the system needs to orphan any "late" microblocks -- in particular, L shouldn't be able to reorganize the chain by sending them late (otherwise they could do shenanigans like selfish mining).  To do this, we can make it so that if the next leader L_next doesn't see any of L's microblocks, L_next commits to L's (on-chain) coinbase microblock as the chain tip. In doing so, L_next forfeits any transaction fees L might have passed on to them, but at least L can't make any ill-gotten gains since its unannounced microblocks are now orphaned.

In case 4, L needs to restart the chain.  That is, L_prev (and possibly previous leaders before it) was in case 2 during its tenures -- it didn't commit, and cannot send any microblocks.  To recover, L must (1) commit to the last active leader's last-seen microblock (which could be its coinbase), and (2) indicate how many previous epochs since then had inactive leaders. L's commitment is only valid if it indicates the exact number of previous inactive leaders; otherwise L's commitment is invalid. If L mines off of a microblock from the last non-empty leader, then L does not have to pay a fork penalty of any kind (discussed later in this STIP). L may mine off of a microblock from an even older non-empty leader, but pays the fork penalty for doing so (discounting inactive epochs).  If L successfully commits, then if L broadcasts its microblocks, we're back to case 1. If L does not broadcast its microblocks, we're back to case 3.

Reply:
Yep -- that is exactly how the leader tenure's should begin -- will add more detail on this to this proposal.